### PR TITLE
refactor: Drop boost::thread and boost::chrono

### PR DESCRIPTION
**Pull request by ken2812221**
Based on #14464 and #14480. Add interruptible thread class to replace boost::thread

The remaining usage of libboost_thread is boost::shared_mutex

### CODE REVIEW

**Comment by ken2812221** [comment]
Apparently I wrote a MSVC-only code. I am going to investigate it.

**Comment by MeshCollider** [comment]
Related #8631

**Comment by ken2812221** [comment]
This is ready for review

File: src/threadinterrupt.cpp [hunk 1]
```diff
@@ -43,3 +43,30 @@ bool CThreadInterrupt::sleep_for(std::chrono::minutes rel_time)
...
+void InterruptionPoint()
+{
+    if (!g_interrupt_flag) { // Not interruptible thread
+        return;
+    }
+    std::lock_guard<std::mutex> lock(g_interrupt_flag->m_mutex);
+    if (g_interrupt_flag->m_interrupted) {
+        throw ThreadInterrupted();
+    }
+}
```
Comment (practicalswift) [inline src/threadinterrupt.cpp]: Missing newline at end of file :-)

**Comment by ken2812221** [comment]
@donaloconnor It should use mutex because it's used by the condition_variable.

For reference, you can find boost implementation for [pthread](https://github.com/boostorg/thread/blob/develop/src/pthread/thread.cpp), it acquires lock before it reads `interrupt_requested`.

**Comment by donaloconnor** [comment]
@ken2812221 ah yes for the sleep. Thanks for the link to boosts impl.

**Comment by ken2812221** [comment]
Actually, I think it's fine to make it use both atomic and mutex at the same time. Then we don't have to lock mutex in InterruptionPoint.

**Comment by donaloconnor** [comment]
In this case it might work but in practice it can cause race conditions. The value of the atomic can change from the moment the cond var is signalled and the predicate is checked (to work with spurious wakeups). Eg: false->true->notify+wakeup->false->predcheck.

The last false flag can never happen here (at least my understanding of the code) I guess so maybe it's okay.

**Comment by ken2812221** [comment]
@donaloconnor I am not sure people would allow this solution. Now the flag contains two variable:
- atomic bool read in `InterruptionPoint`
- non-atomic bool read in `InterruptionSleep`

**Comment by donaloconnor** [comment]
I think it's not worth the complexity tbh. Sorry. I don't know if the perf improvement is worth the extra complexity without measurements. Maybe we should just stick to what boost did as you had in original. I think using the atomic in the cond var is okay for this case because it's state only goes from false->true.
I'd prefer if others chimed in to give feedback on this..
Re: spurious wakeups in the sleep. Do you think that's an issue? The sleep could get ended prematurely. Probably not a big deal. I don't know how much the code relies on accurate sleeps.

**Comment by practicalswift** [comment]
What about splitting the work in two parts?

Start with the low-risk Boost replacement: that change is likely worth it in itself as long as no deviations from what Boost was doing are introduced.

When that work is finished: continue with the risker performance improvement change you are considering.

Does that make sense?

**Comment by MarcoFalke** [comment]
fyi, Boost::Chrono has been removed

**Pull request closed without merging**